src/hotspot/share/opto/divconstants.cpp

```diff
@@ -0,0 +1,209 @@
+// This function does not handle the cases d being a power of 2, which means
+// that c * d is never equal to m.
+//
+// We find the value of c, m such that it satisfies the bounds for both the nonnegative
+// and negative ranges of x. This can be done by finding v_neg and v_pos and the bounds of
+// c * d - m is the intersection of (0, m / v_neg] and (0, m / v_pos). Which is (0, m / v_pos)
+// if v_pos >= v_neg and (0, m / v_neg] otherwise.
+//
+// Given v = max(v_neg, v_pos). The function inductively calculates c, rc, qv, rv such that:
+//
+// c * d - rc = 2**s with 0 < rc <= d
+// qv * v + rv = 2**s with 0 <= rv < v
+template <class T>
+void magic_divide_constants(T d, T N_neg, T N_pos, juint min_s, T& c, bool& c_ovf, juint& s) {
+  static_assert(std::is_unsigned<T>::value, "calculations must be done in the unsigned domain");
+  assert(!is_power_of_2(d), "this case should be handled separately");
+  assert(d <= N_neg || d <= N_pos, "this should just be idealized to 0");
+  constexpr T min_signed = std::numeric_limits<std::make_signed_t<T>>::min();
+  T v_neg = N_neg < d - 1 ? 0 : N_neg - ((N_neg - d + 1) % d);
+  T v_pos = N_pos < d - 1 ? 0 : N_pos - ((N_pos - d + 1) % d);
+  T v = MAX2(v_neg, v_pos);
+  bool tolerate_equal = v_neg > v_pos;
+
+  // base case, s = 0
+  s = 0;
+  c_ovf = false;
+  c = 1;
+  T rc = d - 1;
+  bool qv_ovf = false;
+  T qv = 0;
+  T rv = 1;
+
+  while (true) {
+    if (s >= min_s) {
+      // qv_ovf || rc < qv is the test for rc < qv with overflow taken into
+      // consideration
+      if (qv_ovf || rc < qv || (rc == qv && (tolerate_equal || rv > 0))) {
+        break;
+      }
+    }
+    assert(!c_ovf, "must be");
+    s++;
+
+    if (rc > d - rc) {  // 2 * rc > d
+      c_ovf = c > min_signed;
+      c = c * 2 - 1;
+      rc = rc * 2 - d;
+    } else {            // 2 * rc <= d
+      c_ovf = c >= min_signed;
+      c = c * 2;
+      rc = rc * 2;
+    }
+    if (rv >= v - rv) { // 2 * rv >= v
+      qv_ovf = qv >= min_signed;
+      qv = qv * 2 + 1;
+      rv = rv * 2 - v;
+    } else {            // 2 * rv < v
+      qv_ovf = qv >= min_signed;
+      qv = qv * 2;
+      rv = rv * 2;
+    }
+  }
+}
+template void magic_divide_constants<juint>(juint, juint, juint, juint, juint&, bool&, juint&);
+template void magic_divide_constants<julong>(julong, julong, julong, juint, julong&, bool&, juint&);
+
+// The constant of a W-bit signed division lies in the range of W-bit unsigned
+// integers. As a result, the product of the dividend and the magic constant cannot
+// overflow a 2W-bit signed integer.
+//
+// For unsigned division however, the magic constant may lie outside the range
+// of W-bit unsigned integers, which means the product of it and the dividend can
+// overflow a 2W-bit unsigned integer. In those cases, we use another algorithm.
+//
+// Given s = floor(log2(d)) + W, c = floor(2**s / d) we have
+//
+// floor(x / d) = floor((x + 1) * c / 2**s) for every integer x in [0, 2**W).
+//
```

Comment:
For the record, here's a slightly more general result.

As above, let `N >= 0` be an upper bound for the non-negative dividend `n`, that is, `n` in `[0, N]`, and let divisor `d > 1`.
Further, let
`v = floor(N / d) * d`
The following predicates on real number `x` are equivalent
(a) `(1 / d) * (1 - 1 / (v + 1)) <= x < (1 / d)`
(b) `floor(n / d) = floor((n + 1) * x)` for all `n` in `[0, v + d)`